src/cli/repl.rs

```diff
@@ -39,64 +43,218 @@ impl Validator for InputValidator {
     }
     Ok(())
 }
 
-/// Pads the number of iterations to the first multiple of the reduction count
-/// that's equal or greater than the number of iterations
+/// `pad(a, m)` returns the first multiple of `m` that's equal or greater than `a`
 ///
-/// Panics if reduction count is zero
-#[allow(dead_code)]
-fn pad_iterations(iterations: usize, rc: usize) -> usize {
-    let lower = rc * (iterations / rc);
-    if lower < iterations {
-        lower + rc
+/// Panics if `m` is zero
+#[cfg(not(target_arch = "wasm32"))]
+fn pad(a: usize, m: usize) -> usize {
+    let lower = m * (a / m);
+    if lower < a {
+        lower + m
     } else {
         lower
     }
 }
 
-impl<F: LurkField + serde::Serialize + for<'de> serde::Deserialize<'de>, C: Coprocessor<F>>
-    Repl<F, C>
-{
-    pub fn new(store: Store<F>, env: Ptr<F>, limit: usize, rc: usize) -> Result<Repl<F, C>> {
-        check_non_zero("limit", limit)?;
-        check_non_zero("rc", rc)?;
+#[cfg(not(target_arch = "wasm32"))]
+fn timestamp() -> u128 {
+    use std::time::{SystemTime, UNIX_EPOCH};
+    SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("We're after UNIX_EPOCH")
+        .as_nanos()
+}
+
+type F = pasta_curves::pallas::Scalar; // TODO: generalize this
+
+impl Repl<F> {
+    pub fn new(
+        store: Store<F>,
+        env: Ptr<F>,
+        limit: usize,
+        rc: usize,
+        backend: Backend,
+    ) -> Result<Repl<F>> {
+        validate_non_zero("limit", limit)?;
+        validate_non_zero("rc", rc)?;
+        let field = F::FIELD;
+        backend.validate_field(&field)?;
+        info!("Launching REPL with backend {backend} and field {field}");
         Ok(Repl {
             store,
             env,
             limit,
-            lang: Arc::new(Lang::<F, C>::new()),
-            // last_claim: None,
+            lang: Arc::new(Lang::new()),
             rc,
+            backend,
+            evaluation: None,
         })
     }
 
-    pub fn prove_last_claim(&mut self) -> Result<()> {
-        Ok(())
-        // match &self.last_claim {
-        //     Some(claim) => {
-        //         // TODO
-        //         let _proof = prove_claim(claim);
-        //         Ok(())
-        //     }
-        //     None => {
-        //         bail!("No claim to prove");
-        //     }
-        // }
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn prove_last_frames(&mut self) -> Result<()> {
+        use super::{
+            lurk_proof::LurkProofMeta,
+            paths::{proof_meta_path, proof_path},
+        };
+        use lurk::{
+            proof::{nova::NovaProver, Prover},
+            public_parameters::public_params,
+            z_store::ZStore,
+        };
+        use std::{fs::File, io::BufWriter};
+        match &self.evaluation {
+            None => bail!("No computation to prove"),
+            Some(Evaluation { frames, cost }) => match self.backend {
+                Backend::Nova => {
```

Comment:
This is reimplementing a lot of logic that is already in the `NovaProver` .. is there any chance we could if not reuse it, at least open an issue to make a better use of it?

Reply:
The NovaProver API does not provide the necessary control. It, for example, would redo the evaluation that's already memoized. I will open an issue for it later.

Reply:
The kind of control that I need is too specific. I'm doing some operations (like collecting metadata) in between the separate steps that would happen on a generic function.